src/main/kotlin/calculator/FormulaConverter.kt

```diff
@@ -0,0 +1,83 @@
+package calculator
+
+typealias FormulaConverterValidator = FormulaConverter.Validator
+
+class FormulaConverter {
+    fun convert(formula: String): FormulaElements {
+        formula.shouldValidateFormula()
+
+        val formulas = formula.split(Validator.formulaDelimiter).iterator()
+        return FormulaElements(
+            formulas.next().toDouble(),
+            formulas.asSequence().toList().chunked(2).map {
+                FormulaElement(
+                    type = Operation.of(it[0])!!,
+                    value = it[1].toDouble()
+                )
+            }
+        )
+    }
+
+    private fun String.shouldValidateFormula() = Validator.checkValidateFormulaOrThrows(this)
+
+    object Validator {
+        const val formulaDelimiter = " "
+
+        fun checkValidateFormulaOrThrows(formula: String) {
+            formula.shouldNotEmpty()
+            formula.shouldContainNumber()
+            formula.numbersShouldValidPositions()
+            formula.operationsShouldValidPositions()
+            formula.shouldOddNumberDivided()
+        }
+
+        private fun String.shouldNotEmpty() = require(isNotEmpty())
+
+        private fun String.shouldContainNumber() = require(split().any { it.toDoubleOrNull() != null })
+
+        private fun String.numbersShouldValidPositions() = split().filterIndexed { index, _ -> index.isEven() }
+            .forEach { require(it.toDoubleOrNull() != null) }
+
+        private fun String.operationsShouldValidPositions() = split().filterIndexed { index, _ -> !index.isEven() }
+            .forEach { require(Operation.of(it) != null) }
+
+        private fun String.shouldOddNumberDivided() = require((split().size % 2) == 1)
+
+        private fun String.split() = split(formulaDelimiter)
+
+        private fun Int.isEven() = (this % 2) == 0
+    }
+}
+
+typealias Calculation = (leftOperand: Double, rightOperand: Double) -> Double
+
+enum class Operation(val op: String, val calculation: Calculation) {
+    PLUS(
+        "+",
+        object : Calculation {
+            override fun invoke(leftOperand: Double, rightOperand: Double): Double = leftOperand + rightOperand
+        }
+    ),
```

Comment:
`calculation`이 함수 타입인 만큼 익명 객체가 아닌 람다로 표현하면 더 깔끔해질 것 같아요!
```suggestion
PLUS(
        "+", { leftOperand, rightOperand -> leftOperand + rightOperand }
    ),
```

Reply:
Lambda 를 사용했을때 장점이 더 상세히 표현된 내용이 있어 정리차 포함해봤습니다 :)

- anonymous class 와 비교했을때 lambda 사용했을때 컴파일 후 불필요 class 가 생기지 않음
  - lambda expression 같은경우 'invokedynamic bytecode' 로 치환되므로 속도, 메모리면에서 장점이 있음
- functional interfaces 에서는 람다사용이 일반적
- anonymous class 사용을 위한 불필요 class definition 이 필요없어짐

ref.
https://www.tutorialspoint.com/differences-between-anonymous-class-and-lambda-expression-in-java